Avi's TODO updates

reach2avi opened the pull request:
@katldewitt @thakur-shashank 
not perfect, although most of the TODOs addressed. please share your thoughts. Will further update / modify the rmd based on your feedback. thanks

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,97 @@ model_performance_metrics = function(models, model_names, data){
+  for (idx in 1:n_mod){
+    diagnostics_metrics_df[model_names[idx], "BP_P_Value"] = bptest(models[[idx]])$p.value
+    diagnostics_metrics_df[model_names[idx], "BP_DEC"] = get_bp_decision(models[[idx]])
+    diagnostics_metrics_df[model_names[idx], "SW_P_VALUE"] = shapiro.test(sample(resid(models[[idx]]), 5000))$p.value
```
@reach2avi , please put a comment explaining why you're sample 5000 observations here. This is a limitation you will need to discuss below because of the size our dataset.

reach2avi replied:
@katldewitt good point. thought of including it last night itself but then forgot as it was late.

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,97 @@ model_performance_metrics = function(models, model_names, data){
+  )
+  
+  for (idx in 1:n_mod){
+    diagnostics_metrics_df[model_names[idx], "BP_P_Value"] = bptest(models[[idx]])$p.value
```
The p-values for both tests are rendering as 0's. Please use `format.pval()` to show the true values of these statistics

reach2avi replied:
done @katldewitt

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,97 @@ model_performance_metrics = function(models, model_names, data){
+}
+
+# function to get decision of Breusch-Pagan Test (for checking homoscedasticity)
+get_bp_decision = function(model, alpha=0.05) {
```
Right now, your driver function calculates BP twice. Consider modifying this function to return BOTH the p-value and the decision to reduce the unnecessary calls; same with ShapiroTest. In the Shapiro test, there's also a risk that the P value you're reporting is from a different sample than the decision you're making. Either use consistent obs [1:5000] or only sample 1 time.

reach2avi replied:
@katldewitt done.

reach2avi replied:
@katldewitt completed the first part. 
with the 2nd part which is a great point - would setting seed inside the function help?

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,97 @@ model_performance_metrics = function(models, model_names, data){
+
+*models      : * A list of models and model names to generate the results table
+
+**Returns    : ** None
```
Please change this to "Outputs" and describe the resulting table. What do rows represent? What columns will appear? See Shashank's description above as an example.

reach2avi replied:
@katldewitt  done

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,97 @@ model_performance_metrics = function(models, model_names, data){
+
+*models      : * A list of models to generate the plots
+
+**Returns    : ** None
```
Same comment as above. Describe the expected output here.

reach2avi replied:
@katldewitt done

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,97 @@ model_performance_metrics = function(models, model_names, data){
+}
+
+# function to plot fitted vs. residual plot to test homoscedasticity - uniform variation in error plotted against the fitted value or response  
+plot_fitted_resid = function(model, model_name, stitle) {
```
Please either a) use the parameter `stitle` or 2) remove unused  parameter `stitle`

reach2avi replied:
@katldewitt  done

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
````diff
@@ -524,6 +620,35 @@ plot(final_selected_model,
      col = green_color)
 ```
 
+We also ran model diagnostic tests to check linear regression assumptions with the help of Breusch-Pagan Test and Shapiro-Wilk Test. 
````
1) Please remove the "- " since this should be written as a narrative.
2) Please explicitly stat that "The constant variance assumption is violated." for the BP Test results
3) There's inconsistent formatting around quotes here ('Reject the null hypothesis', "Reject the null hypothesis')
4) Spelling error "al" -> "all"
5) State what alpha you are using (in this case, doesn't really matter since we'll reject regardless)
6) I think the conclusion from the Shapiro-Wilk test is not "model error is not following a normal distribution.", but rather that the data used in the model are not drawn from a normal distribution. https://book.stat420.org/model-diagnostics.html#shapiro-wilk-test

reach2avi replied:
@katldewitt done

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
````diff
@@ -524,6 +620,35 @@ plot(final_selected_model,
+get_model_diagnostic_tests(cahoudta_models, cahoudta_model_names)
+```
+
+We are showing the model diagnostic plots for our top 2 models 1) cahoudta_full_add_model and 2) cahoudta_add_bckwd_bic_model. As it can be see, with log transformation of response we could reduce the issue of heteroscedasticity and error not following normal distribution. However, this is not entirely gone as it can be seen from the plots below. 
````
1. Typo "As it can be see," --> "As it can be seen"
2. Formatting - please use ` around model names for consistency i.e. `cahoudta_full_add_model` and `cahoudta_add_bckwd_bic_model`
3. If you'd like to show the improvements from log to non-log, it may be worth adding a graph of the `cahoudta_add_bckwd_bic_model`'s parameters against the NON-logged response next to the `cahoudta_add_bckwd_bic_model` model

reach2avi replied:
@katldewitt done 1 & 2. 
feel like another set of chart may hinder with framework / storyline. so not adding pre-log- transformation plots.

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
````diff
@@ -524,6 +620,35 @@ plot(final_selected_model,
+```
+
 
 <!TODO: Avi's helper methods related to results should go under "METHODS" Section. Call the driver function here.>
````
Please remove TODO's as you complete them.

reach2avi replied:
@katldewitt done

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
 
 It is also important to underscore future directions for analyses on this data set that our group was not able to implement due to time constraints or the focus of the material of STAT420. This section is meant to emphasize the careful consideration our team took into ideating about the best way to model the data.
 
+- Extracting meaningful Geo-Spatial insights / predictor variables based on 'Latitude' & 'Longitude' data: Houses within 50 miles radius of largest cities in California would likely have higher prices compared to houses which are outside 50 miles radius. Our team ideated extracting variables using Geo-Spatial libraries and methods such as  'Haversine Distance' - formula to calculate distance betw ...
```
1) This section should be written as a narrative and not a bulleted list. Please include a topic sentence for each bullet instead of a fragment.
2) please use ` for variable names
3) Inconsistency of "lat"/"latitude" and "long"/"longitude", please use the full variable name throughout the report.
4) " 'Haversine Distance' - formula " --> remove extra spaces and " 'Haversine Distance', a formula used..."
5) Write out 2 as "two"
6) " In the interest of time we could not create these variables but we certainly would like to try including these Geo-Spatial variables in our model. " --> " In the interest of time, we could not create these variables, but we believe a future direction with this data is including these Geo-Spatial variables in modeling."

reach2avi replied:
@katldewitt resolved. updated verbiage / language

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
 
+- Extracting meaningful Geo-Spatial insights / predictor variables based on 'Latitude' & 'Longitude' data: Houses within 50 miles radius of largest cities in California would likely have higher prices compared to houses which are outside 50 miles radius. Our team ideated extracting variables using Geo-Spatial libraries and methods such as  'Haversine Distance' - formula to calculate distance betw ...
+
+- Looking at the response variable histogram, it seems that the median house prices are capped at the top. We would like to further explore the impacts of this in response variable. or maybe limit the model usage for houses with very high value 
```
This paragraph ends in a fragment. Let's either remove the fragment or flesh out the thought. I don't understand the concrete next step to this suggestion from a statistical perspective (are you suggesting imputation of data? Truncating responses?). To me, this paragraph should be considered a limitation, and not a future direction.

reach2avi replied:
@katldewitt 
yes. suggesting to truncate and model. due to capping of response the model process will not be accurate  anyway in this rnage. its not a missing value issue so not recommending imputation.

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
+
+- Looking at the response variable histogram, it seems that the median house prices are capped at the top. We would like to further explore the impacts of this in response variable. or maybe limit the model usage for houses with very high value 
+
+- Our 9th model - cahoudta_2int_txfd_model - has the highest adjusted r-square although it suffers from over-fitting issue. We would like to investigate this further with the objective of mitigating over-fitting withough sacrificing adjusted r-square performance. 
```
@reach2avi , this model is very overfit as is visible based on the high test RMSE. I do not understand what this future direction is proposing since the test RMSE was our deciding factor. We should NOT be mentioning this in future directions.

cc: @thakur-shashank

reach2avi replied:
@katldewitt 

I want to say that over test data we got a model with better r-sq. however, rightfully, we are not selecting it as one of the top 2 models. as a next step it may help to further investigate why the performance is not good in the test data and incorporate learnings (if any). felt worth looking into as a next step.

katldewitt replied:
Hey @reach2avi ,

I recommend reading these articles about the dangers of overfitting as it relates to p-hacking.

1. Pitfall # 3: https://www.capitalone.com/tech/machine-learning/10-common-machine-learning-mistakes/
2. https://gregpark.io/blog/Kaggle-Psychopathy-Postmortem/
3. https://datascience.stanford.edu/news/data-snooping

cc: @thakur-shashank

thakur-shashank replied:
@katldewitt and @reach2avi I read through all the articles. We should explain why we didn't select the model because it is overfitting and stop at that, rather than trying to draw any conclusion from partial results of just r-squared value.

reach2avi replied:
@thakur-shashank @katldewitt 
let me explain what i am conveying: 
1. the mode is doing good over train but not on test. its overfitting 
2. THATS WHY ITS NOT IN TOP 2 
3. although, want to investigate further as to why as a next step. 
4. not advocating to include this model or anything else. but i am curious as to why its performing poorly in test. and want to investigate as a next step. 
5.

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
+
+- Our 9th model - cahoudta_2int_txfd_model - has the highest adjusted r-square although it suffers from over-fitting issue. We would like to investigate this further with the objective of mitigating over-fitting withough sacrificing adjusted r-square performance. 
+
+- In our dataset 'total_bed_rooms' column had missing values. Given this column was highly correlated with 'total_bedrooms' and number of missing value being low - we safely deleted 'total_bedrooms' variable from the modeling exercise. In future, we would like to treat 'total_bedrooms' with missing value imputation to find out whether it has stronger signal than 'total_rooms' variable 
```
1) Grammar/formatting "Given this column was highly correlated with 'total_bedrooms' and number of missing value being low - we " --> "Given this column was highly correlated with `total_bedrooms` and number of missing value being **low,** we "
2) Missing period at the end of the paragraph.
3) Please format ALL variable names with `.

reach2avi replied:
@katldewitt  
not a stong point as most of the predictive power of bedrooms var is getting captured by rooms var. so deleting this point. thught of keeping it earlier to show possibility of missing value imputation in the modeling process as we did not get a chance to do it in our project

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
+
+- In our dataset 'total_bed_rooms' column had missing values. Given this column was highly correlated with 'total_bedrooms' and number of missing value being low - we safely deleted 'total_bedrooms' variable from the modeling exercise. In future, we would like to treat 'total_bedrooms' with missing value imputation to find out whether it has stronger signal than 'total_rooms' variable 
+
+- We see some from BP Test result and 'Fitted vs. Residual' plot that the residuals still have some heteroscedasticity. In future, we would like to further reduce the issue by trying tranformation of predictor varibales 
```
1) spelling "varibales " --> "variables"
2) missing period at the end.

reach2avi replied:
@katldewitt done

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
+
+- We see some from BP Test result and 'Fitted vs. Residual' plot that the residuals still have some heteroscedasticity. In future, we would like to further reduce the issue by trying tranformation of predictor varibales 
+
+- Our Q-Q plot for top 2 models does mostly look good, however towards the lower edge we see some diversion from the line. Time permitting, we would investigate more going forward to resolve the normality of error term issue 
```
Didn't we reject the tests for both BP and Shapiro? If that's the case, we shouldn't say QQPlots look good.

reach2avi replied:
@kathryn 
here what i wanted to convey was that these two issues have gotten much better due to steps we have taken - for ex: transforming response variable. but still issues remain as both tests are failing. meant to say Q-Q plot looks better compared to when run over raw data. to highlight lessons learnt from the course and showing its impact. although the test still fails there has been progress. time permitting a future step could be to further optimize transformation to reduce the impact of heteroscedasticity and normality of error. changing wording a bit to highlight above 

updated text below 
================

We see some from BP Test result and 'Fitted vs. Residual' plot that the residuals still have some heteroscedasticity. In future, we would like to further reduce the issue by trying tranformations of predictor variables  

- Our Q-Q plot for top 2 models looks much better compared to when we started modeling with the raw data.Time permitting, we would investigate more as to how to further mitigate the normality of error issue we see in the data.

reach2avi replied:
@katldewitt updated text to explain the point. some ideas of things that we can further try and possibly improve upon.

katldewitt commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -570,6 +695,18 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
+
+- Our Q-Q plot for top 2 models does mostly look good, however towards the lower edge we see some diversion from the line. Time permitting, we would investigate more going forward to resolve the normality of error term issue 
+
 <!TODO: Avi to add details around thought process of using variable transformations such as converting lat/lon to
```
Please remove the <!TODO's

reach2avi replied:
@katldewitt done

katldewitt reviewed (changes requested):
Hey @reach2avi , 

I reviewed your changes and have several comments. Please review and let me know if you have questions.

Thanks!
cc: @thakur-shashank

reach2avi commented on 200_Final_Report/200-final-report.Rmd:
````diff
@@ -410,7 +502,11 @@ message("Model 8: Formula: ", less_preds)
 ```
 
-Our final step was to attempt variable transformations to attempt to better capture the variance in the response variable. We transformed predictors `population` and `total_rooms` to be normalized with predictor `households`. <!TODO: Avi, can you add about your reasoning here why we need additional variables transformed. The results are showing that this model is over-fitting?>
+Our final step was to attempt variable transformations to attempt to better capture the variance in the response variable. We transformed predictors `population` and `total_rooms` to be normalized with predictor `households`. We think derived variables 'rooms_per_hh' and 'pop_per_hh' should add predictive power to the model because of the following reasons. 
````
updated @katldewitt

reach2avi commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -410,7 +502,11 @@ message("Model 8: Formula: ", less_preds)
 
-Our final step was to attempt variable transformations to attempt to better capture the variance in the response variable. We transformed predictors `population` and `total_rooms` to be normalized with predictor `households`. <!TODO: Avi, can you add about your reasoning here why we need additional variables transformed. The results are showing that this model is over-fitting?>
+Our final step was to attempt variable transformations to attempt to better capture the variance in the response variable. We transformed predictors `population` and `total_rooms` to be normalized with predictor `households`. We think derived variables 'rooms_per_hh' and 'pop_per_hh' should add predictive power to the model because of the following reasons. 
+- 'rooms_per_hh': houses with higher number of rooms per household is likely going to be larger having a positive correlation with housing prices 
```
done. @katldewitt

reach2avi commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -524,6 +620,35 @@ plot(final_selected_model,
      col = green_color)
```
done @katldewitt

reach2avi commented:
@katldewitt good feedback. thank a lot. addressed the comments and sending for another ronud

thakur-shashank commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,112 @@ model_performance_metrics = function(models, model_names, data){
+  }
+}
+
+# function to plot fitted vs. residual plot to test homoscedasticity - uniform variation in error plotted against the fitted value or response  
```
For sake of consistency add the function documentation similar to the `get_model_diagnostic_plots`. Or combine with the main driver function, if you're not using these functions independently.

thakur-shashank commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,112 @@ model_performance_metrics = function(models, model_names, data){
+}
+
+# function to plot fitted vs. residual plot to test homoscedasticity - uniform variation in error plotted against the fitted value or response  
+plot_fitted_resid = function(model, model_name) {
```
For sake of consistency add the function documentation similar to the `get_model_diagnostic_plots`. Or combine with the main driver function, if you're not using these functions independently.

thakur-shashank replied:
Also I have formatted the code of function, so when we print PDF, the code doesn't overflow the coding box.

reach2avi replied:
@thakur-shashank thanks for the formatting the pdf. for modularity of code kept small independent modular function. they also have a a bit of comment to explain what they do. as they are being used in the main helped. they are part of the same code block too. but if you think otherwise more than happy to either combine or write full help documentation for the smaller function. let me know.

reach2avi replied:
@thakur-shashank this is resolved include in the helper function code.

thakur-shashank commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -119,7 +121,112 @@ model_performance_metrics = function(models, model_names, data){
+}
+
+# function to plot qq-plot & qq-line
+plot_qq = function(model, model_name) {
```
For sake of consistency add the function documentation similar to the `get_model_diagnostic_plots`. Or combine with the main driver function, if you're not using these functions independently.

reach2avi replied:
@thakur-shashank 
on second thoughts let me combine wiht the main helper function. and keep things consistent. will do

thakur-shashank commented on 200_Final_Report/200-final-report.Rmd:
```diff
@@ -695,22 +697,14 @@ The other key $\hat{\beta}$ to call attention to is that for `median_income`. Fo
+Our team discussed about the potential of extracting meaningful Geo-Spatial insights / predictor variables based on 'Latitude' & 'Longitude' data. The hypothesis being, houses within 50 miles radius of large cities in California would likely have higher median prices. We considered leveraging Geo-Spatial libraries for this purpose. For example: 'Haversine Formula' to calculate distance between tw ...
 
-- In our dataset 'total_bed_rooms' column had missing values. Given this column was highly correlated with 'total_bedrooms' and number of missing value being low - we safely deleted 'total_bedrooms' variable from the modeling exercise. In future, we would like to treat 'total_bedrooms' with missing value imputation to find out whether it has stronger signal than 'total_rooms' variable 
+Looking at the response variable histogram, we found that median house prices are capped at the top. We would like to further explore the impacts of this in our models from a fit perspective - possibly truncating observations with very high median house prices. Due to capping of the response variable, the accuracy of the model in this price range would not be good anyway.  
```
@reach2avi I tired to reword this section. But realized I don't understand what you mean by "...median house prices are capped at the top..." and subsequent analysis. Especially I was lost at the statement..."..accuracy of the model in this price range would not be good anyway.." Can you please expand your explanation or drop this....

katldewitt replied:
My vote is we remove this section to be more concise about our next steps/future directions.

reach2avi replied:
@thakur-shashank 
i will make an attempt to explain this further. the explanation needs a little bit of work. 

the point is if you look at the histogram of the response var from the EDA section - you would see an unexpected higher bar on the right of the histogram. instead of a long tail. this behavior is very consistent with data capture / capping issue. Higher price houses are capped at fix value range - thats why we dont see a long tail. given this issues is there with response var itself - prediction / modeling for houses in this price range is not going to be accurate. 

let me know if  the above explanation / observation makes sense

thakur-shashank reviewed (changes requested):
@reach2avi I have added my review comments. In interest of time I made some changes to the RMD file and pushed them to your branch. Please review and let me know if you've questions.

@katldewitt I made few minor adjustments to discussion and interpretation of model sections.

katldewitt commented:
> @reach2avi I have added my review comments. In interest of time I made some changes to the RMD file and pushed them to your branch. Please review and let me know if you've questions.
> 
> @katldewitt I made few minor adjustments to discussion and interpretation of model sections.

Thanks @thakur-shashank and @reach2avi for all your work today! I pulled Shashank's updates, reviewed Avi's changes, and reknit. When I read from end to end, I noticed a few things that could use tightening up. I'm going to approve on the condition that Shashank's feedback about helper functions get incorporated. After those changes, I trust you both to submit.

reach2avi commented:
@thakur-shashank 
addressed all comments. 
took me a couple of hours to resolve merge conflict. please check all the changes you made today afternoon are included

thakur-shashank reviewed (approved):
I have completed my review. Some of my edits from yesterday were wiped out. I have made those edits again and pushed changes to feat/final/avi. Let just compile and review one final time only for grammatical mistakes, spellings and cosmetics..
After @reach2avi generates the HTML document for final review, I think we should. submit it...
 
 CC: @katldewitt
